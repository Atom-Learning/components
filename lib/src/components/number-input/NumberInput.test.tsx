import { TooltipProvider } from '@radix-ui/react-tooltip'
import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'
import * as React from 'react'

import { NumberInput } from '.'
import type { NumberInputProps } from './NumberInput'

const renderComponent = (props: Partial<NumberInputProps> = {}) => {
  return render(
    <TooltipProvider>
      <NumberInput name="test" aria-label="test" {...props} />
    </TooltipProvider>
  )
}

describe(`NumberInput component`, () => {
  it('renders number input ', async () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('has no programmatically detectable a11y issues', async () => {
    const { container } = renderComponent()
    expect(await axe(container)).toHaveNoViolations()
  })

  it('should start with defaultValue of 0', async () => {
    renderComponent()
    expect(screen.getByRole('spinbutton')).toHaveValue('0')
  })

  it('should increment value when incremenent button is pressed', async () => {
    renderComponent()

    const input = screen.getByRole('spinbutton')
    const incrementBtn = screen.getByRole('button', { name: /increment/i })

    await userEvent.click(incrementBtn)
    expect(input).toHaveValue('1')

    await userEvent.click(incrementBtn)
    expect(input).toHaveValue('2')
  })

  it('should decrement value when decrement button is pressed', async () => {
    renderComponent({ defaultValue: 5 })

    const input = screen.getByRole('spinbutton')
    const decrementBtn = screen.getByRole('button', { name: /decrement/i })

    await userEvent.click(decrementBtn)
    expect(input).toHaveValue('4')

    await userEvent.click(decrementBtn)
    expect(input).toHaveValue('3')
  })

  it('should disable increment/decrement buttons if at min/max values', () => {
    renderComponent({ max: 10 })

    const input = screen.getByRole('spinbutton')
    const decrementBtn = screen.getByRole('button', { name: /decrement/i })
    const incrementBtn = screen.getByRole('button', { name: /increment/i })

    expect(decrementBtn).toBeDisabled()

    fireEvent.keyDown(input, { key: 'End' })
    expect(incrementBtn).toBeDisabled()
  })

  it('should show tooltip on disabled button hover', async () => {
    renderComponent()
    fireEvent.focus(screen.getByRole('button', { name: /decrement/i }))
    expect(await screen.findByRole('tooltip')).toBeInTheDocument()
  })

  it('should increment/decrement on keyboard interaction', () => {
    renderComponent({ max: 10 })

    const input = screen.getByRole('spinbutton')
    input.focus()

    fireEvent.keyDown(input, { key: 'ArrowUp' })
    expect(input).toHaveValue('1')

    fireEvent.keyDown(input, { key: 'ArrowDown' })
    expect(input).toHaveValue('0')

    fireEvent.keyDown(input, { key: 'End' })
    expect(input).toHaveValue('10')

    fireEvent.keyDown(input, { key: 'Home' })
    expect(input).toHaveValue('0')
  })

  it('should call onValueChange prop when value changes', async () => {
    const onValueChange = jest.fn()
    renderComponent({ onValueChange })

    const incrementBtn = screen.getByRole('button', { name: /increment/i })
    await userEvent.click(incrementBtn)

    expect(onValueChange).toBeCalled()
    expect(onValueChange).toBeCalledWith(1)
  })

  it('renders lg size', async () => {
    const { container } = renderComponent({ size: 'lg' })

    expect(container).toMatchSnapshot()
  })
})
