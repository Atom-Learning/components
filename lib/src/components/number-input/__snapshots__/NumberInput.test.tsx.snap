// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput component renders lg size 1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-jVLVTy {
    z-index: 1;
  }

  .c-oUmPa {
    align-items: center;
    -webkit-appearance: none;
    appearance: none;
    background: white;
    border: unset;
    border-radius: var(--radii-0);
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    flex-shrink: 0;
    justify-content: center;
    padding: unset;
    transition: all 100ms ease-out;
  }

  .c-oUmPa[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-hMRxhp {
    display: inline-block;
    fill: none;
    flex-shrink: 0;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-dWvLOH {
    border-color: var(--colors-tonal400) !important;
    z-index: 1;
    height: 100% !important;
  }

  .c-dWvLOH:hover {
    background: var(--colors-tonal50);
  }

  .c-dWvLOH:hover svg {
    color: var(--colors-tonal400);
  }

  .c-dWvLOH svg {
    color: var(--colors-tonal300);
  }

  .c-dWvLOH:active {
    background: var(--colors-tonal100);
  }

  .c-dWvLOH:active svg {
    color: var(--colors-tonal500);
  }

  .c-dWvLOH:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-hVoQtf {
    background-color: var(--colors-tonal500);
    border-radius: var(--radii-0);
    box-shadow: var(--shadows-0);
    color: white;
    font-family: var(--fonts-body);
    font-size: var(--fontSizes-sm);
    line-height: 1.5;
    white-space: normal;
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    padding-top: var(--space-2);
    padding-bottom: var(--space-2);
    z-index: 10;
  }

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf {
      animation-duration: 75ms;
      animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);
      will-change: transform, opacity;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="top"] {
      animation-name: k-ftcNPK;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="right"] {
      animation-name: k-hbaHED;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="bottom"] {
      animation-name: k-daNevH;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="left"] {
      animation-name: k-fvyGIa;
    }
}

  .c-lfQsGQ {
    -webkit-appearance: none;
    appearance: none;
    border: 1px solid var(--colors-tonal400);
    border-radius: var(--radii-0);
    box-shadow: none;
    box-sizing: border-box;
    color: var(--colors-tonal600);
    cursor: text;
    display: block;
    font-family: var(--fonts-body);
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    transition: all 100ms ease-out;
    width: 100%;
  }

  .c-lfQsGQ:focus {
    border-color: var(--colors-primary800);
    outline: none;
  }

  .c-lfQsGQ[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-lfQsGQ::placeholder {
    color: var(--colors-tonal300);
    opacity: 1;
  }

  .c-kfLCww {
    fill: var(--colors-tonal500);
  }

  [data-align="end"] .c-kfLCww {
    margin-right: var(--space-2);
  }

  [data-align="start"] .c-kfLCww {
    margin-left: var(--space-2);
  }
}

@media  {
  .c-oUmPa-PrXKS-size-md {
    height: var(--sizes-4);
    width: var(--sizes-4);
  }

  .c-hMRxhp-dMrjnF-size-md {
    height: var(--sizes-2);
    width: var(--sizes-2);
    stroke-width: 1.75;
  }

  .c-hVoQtf-gWQnqe-size-md {
    max-width: 250px;
  }

  .c-lfQsGQ-fwoCmE-size-md {
    height: var(--sizes-4);
    font-size: var(--fontSizes-md);
    line-height: 2;
  }

  .c-lfQsGQ-jAZHIx-size-lg {
    height: var(--sizes-5);
    font-size: var(--fontSizes-md);
    line-height: 2;
  }
}

@media  {
  .c-oUmPa-fboUUB-cv {
    border: 1px solid;
    border-color: currentColor;
    color: var(--colors-primary800);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):hover,
  .c-oUmPa-fboUUB-cv:not(:disabled):focus {
    color: var(--colors-primary900);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):active {
    color: var(--colors-primary1000);
  }
}

@media  {
  .c-dWvLOH-iwWtoZ-css {
    border-right: none;
    border-top-right-radius: 0px;
    border-bottom-right-radius: 0px;
  }

  .c-lfQsGQ-ibwGleD-css {
    border-radius: 0px;
    width: var(--sizes-6);
  }

  .c-lfQsGQ-ibwGleD-css:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-dWvLOH-ihgsOSq-css {
    border-left: none;
    border-top-left-radius: 0px;
    border-bottom-left-radius: 0px;
  }
}

<div>
  <div
    class="c-dhzjXW"
  >
    <span
      class="c-jVLVTy c-PJLV"
      data-state="closed"
      tabindex="-1"
    >
      <button
        aria-label="decrement"
        class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-iwWtoZ-css"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 12H4"
          />
        </svg>
      </button>
    </span>
    <input
      aria-label="test"
      aria-valuemax="9007199254740991"
      aria-valuemin="0"
      aria-valuenow="0"
      class="c-lfQsGQ c-lfQsGQ-jAZHIx-size-lg c-lfQsGQ-ibwGleD-css"
      inputmode="numeric"
      name="test"
      pattern="[0-9]*"
      role="spinbutton"
      type="text"
      value="0"
    />
    <span
      class="c-jVLVTy c-PJLV"
      data-state="closed"
      tabindex="-1"
    >
      <button
        aria-label="increment"
        class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-ihgsOSq-css"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 12H4m8-8v16"
          />
        </svg>
      </button>
    </span>
  </div>
</div>
`;

exports[`NumberInput component renders number input  1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-jVLVTy {
    z-index: 1;
  }

  .c-oUmPa {
    align-items: center;
    -webkit-appearance: none;
    appearance: none;
    background: white;
    border: unset;
    border-radius: var(--radii-0);
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    flex-shrink: 0;
    justify-content: center;
    padding: unset;
    transition: all 100ms ease-out;
  }

  .c-oUmPa[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-hMRxhp {
    display: inline-block;
    fill: none;
    flex-shrink: 0;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-dWvLOH {
    border-color: var(--colors-tonal400) !important;
    z-index: 1;
    height: 100% !important;
  }

  .c-dWvLOH:hover {
    background: var(--colors-tonal50);
  }

  .c-dWvLOH:hover svg {
    color: var(--colors-tonal400);
  }

  .c-dWvLOH svg {
    color: var(--colors-tonal300);
  }

  .c-dWvLOH:active {
    background: var(--colors-tonal100);
  }

  .c-dWvLOH:active svg {
    color: var(--colors-tonal500);
  }

  .c-dWvLOH:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-hVoQtf {
    background-color: var(--colors-tonal500);
    border-radius: var(--radii-0);
    box-shadow: var(--shadows-0);
    color: white;
    font-family: var(--fonts-body);
    font-size: var(--fontSizes-sm);
    line-height: 1.5;
    white-space: normal;
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    padding-top: var(--space-2);
    padding-bottom: var(--space-2);
    z-index: 10;
  }

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf {
      animation-duration: 75ms;
      animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);
      will-change: transform, opacity;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="top"] {
      animation-name: k-ftcNPK;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="right"] {
      animation-name: k-hbaHED;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="bottom"] {
      animation-name: k-daNevH;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="left"] {
      animation-name: k-fvyGIa;
    }
}

  .c-lfQsGQ {
    -webkit-appearance: none;
    appearance: none;
    border: 1px solid var(--colors-tonal400);
    border-radius: var(--radii-0);
    box-shadow: none;
    box-sizing: border-box;
    color: var(--colors-tonal600);
    cursor: text;
    display: block;
    font-family: var(--fonts-body);
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    transition: all 100ms ease-out;
    width: 100%;
  }

  .c-lfQsGQ:focus {
    border-color: var(--colors-primary800);
    outline: none;
  }

  .c-lfQsGQ[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-lfQsGQ::placeholder {
    color: var(--colors-tonal300);
    opacity: 1;
  }
}

@media  {
  .c-oUmPa-PrXKS-size-md {
    height: var(--sizes-4);
    width: var(--sizes-4);
  }

  .c-hMRxhp-dMrjnF-size-md {
    height: var(--sizes-2);
    width: var(--sizes-2);
    stroke-width: 1.75;
  }

  .c-hVoQtf-gWQnqe-size-md {
    max-width: 250px;
  }

  .c-lfQsGQ-fwoCmE-size-md {
    height: var(--sizes-4);
    font-size: var(--fontSizes-md);
    line-height: 2;
  }
}

@media  {
  .c-oUmPa-fboUUB-cv {
    border: 1px solid;
    border-color: currentColor;
    color: var(--colors-primary800);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):hover,
  .c-oUmPa-fboUUB-cv:not(:disabled):focus {
    color: var(--colors-primary900);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):active {
    color: var(--colors-primary1000);
  }
}

@media  {
  .c-dWvLOH-iwWtoZ-css {
    border-right: none;
    border-top-right-radius: 0px;
    border-bottom-right-radius: 0px;
  }

  .c-lfQsGQ-ibwGleD-css {
    border-radius: 0px;
    width: var(--sizes-6);
  }

  .c-lfQsGQ-ibwGleD-css:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-dWvLOH-ihgsOSq-css {
    border-left: none;
    border-top-left-radius: 0px;
    border-bottom-left-radius: 0px;
  }
}

<div>
  <div
    class="c-dhzjXW"
  >
    <span
      class="c-jVLVTy c-PJLV"
      data-state="closed"
      tabindex="-1"
    >
      <button
        aria-label="decrement"
        class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-iwWtoZ-css"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 12H4"
          />
        </svg>
      </button>
    </span>
    <input
      aria-label="test"
      aria-valuemax="9007199254740991"
      aria-valuemin="0"
      aria-valuenow="0"
      class="c-lfQsGQ c-lfQsGQ-fwoCmE-size-md c-lfQsGQ-ibwGleD-css"
      inputmode="numeric"
      name="test"
      pattern="[0-9]*"
      role="spinbutton"
      type="text"
      value="0"
    />
    <span
      class="c-jVLVTy c-PJLV"
      data-state="closed"
      tabindex="-1"
    >
      <button
        aria-label="increment"
        class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-ihgsOSq-css"
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 12H4m8-8v16"
          />
        </svg>
      </button>
    </span>
  </div>
</div>
`;
